controller/api/proxy/fallback_profile_listener.go

```diff
@@ -0,0 +1,114 @@
+package proxy
+
+import (
+	"sync"
+
+	sp "github.com/linkerd/linkerd2/controller/gen/apis/serviceprofile/v1alpha1"
+)
+
+type fallbackProfileListener struct {
+	underlying profileUpdateListener
+	primary    *primaryProfileListener
+	backup     *backupProfileListener
+	mutex      sync.Mutex
+}
+
+type primaryProfileListener struct {
+	state  *sp.ServiceProfile
+	parent *fallbackProfileListener
+}
+
+type backupProfileListener struct {
```

Comment:
thanks for implementing this, a lot clearer!

tioli, you could avoid duplicating the implementations between `primaryProfileListener` and `backupProfileListener` by implementing a shared `parentListener`:

```diff
diff --git a/controller/api/proxy/fallback_profile_listener.go b/controller/api/proxy/fallback_profile_listener.go
index 8e29d99e..31c0d450 100644
--- a/controller/api/proxy/fallback_profile_listener.go
+++ b/controller/api/proxy/fallback_profile_listener.go
@@ -13,14 +13,18 @@ type fallbackProfileListener struct {
 	mutex      sync.Mutex
 }
 
-type primaryProfileListener struct {
-	state  *sp.ServiceProfile
+type parentListener struct {
 	parent *fallbackProfileListener
 }
 
+type primaryProfileListener struct {
+	parentListener
+	state *sp.ServiceProfile
+}
+
 type backupProfileListener struct {
-	state  *sp.ServiceProfile
-	parent *fallbackProfileListener
+	parentListener
+	state *sp.ServiceProfile
 }
 
 // newFallbackProfileListener takes an underlying profileUpdateListener and
@@ -36,10 +40,10 @@ func newFallbackProfileListener(listener profileUpdateListener) (profileUpdateLi
 	}
 
 	primary := primaryProfileListener{
-		parent: &fallback,
+		parentListener: parentListener{parent: &fallback},
 	}
 	backup := backupProfileListener{
-		parent: &fallback,
+		parentListener: parentListener{parent: &fallback},
 	}
 	fallback.primary = &primary
 	fallback.backup = &backup
@@ -68,18 +72,6 @@ func (p *primaryProfileListener) Update(profile *sp.ServiceProfile) {
 	p.parent.underlying.Update(nil)
 }
 
-func (p *primaryProfileListener) ClientClose() <-chan struct{} {
-	return p.parent.underlying.ClientClose()
-}
-
-func (p *primaryProfileListener) ServerClose() <-chan struct{} {
-	return p.parent.underlying.ServerClose()
-}
-
-func (p *primaryProfileListener) Stop() {
-	p.parent.underlying.Stop()
-}
-
 // Backup
 
 func (b *backupProfileListener) Update(profile *sp.ServiceProfile) {
@@ -101,14 +93,16 @@ func (b *backupProfileListener) Update(profile *sp.ServiceProfile) {
 	b.parent.underlying.Update(nil)
 }
 
-func (b *backupProfileListener) ClientClose() <-chan struct{} {
-	return b.parent.underlying.ClientClose()
+// Parent
+
+func (p *parentListener) ClientClose() <-chan struct{} {
+	return p.parent.underlying.ClientClose()
 }
 
-func (b *backupProfileListener) ServerClose() <-chan struct{} {
-	return b.parent.underlying.ServerClose()
+func (p *parentListener) ServerClose() <-chan struct{} {
+	return p.parent.underlying.ServerClose()
 }
 
-func (b *backupProfileListener) Stop() {
-	b.parent.underlying.Stop()
+func (p *parentListener) Stop() {
+	p.parent.underlying.Stop()
 }
```